crash_rate_aggregates/fill_database.py

```diff
@@ -0,0 +1,222 @@
+
+        subprocess_crash_plugin = get_property(ping, ("payload", "keyedHistograms", "SUBPROCESS_CRASHES_WITH_DUMP", "plugin"))
+        result["subprocess_crash_plugin"] = 0 if subprocess_crash_plugin is None else Histogram("SUBPROCESS_CRASHES_WITH_DUMP", subprocess_crash_plugin).get_value()
+
+        subprocess_crash_gmplugin = get_property(ping, ("payload", "keyedHistograms", "SUBPROCESS_CRASHES_WITH_DUMP", "gmplugin"))
+        result["subprocess_crash_gmplugin"] = 0 if subprocess_crash_gmplugin is None else Histogram("SUBPROCESS_CRASHES_WITH_DUMP", subprocess_crash_gmplugin).get_value()
+        return result
+
+    def is_valid(ping): # sanity check to make sure the ping is actually usable for our purposes
+        if not isinstance(ping["submission_date"], date):
+            return False
+        if not isinstance(ping["activity_date"], date):
+            return False
+        subsession_length = ping["subsession_length"]
+        if subsession_length is not None and not isinstance(subsession_length, (int, long)):
+            return False
+        if not isinstance(ping["build_id"], (str, unicode)):
+            return False
+        if not re.match(r"^\d{14}$", ping["build_id"]): # only accept valid buildids
+            return False
+        return True
+
+    original_count = pings.count()
+    ping_properties = pings.map(get_properties).filter(is_valid).cache()
+    filtered_count = ping_properties.count()
+
+    def get_crash_pair(ping): # responsible for normalizing a single ping into a crash pair
+        submission_date = ping["submission_date"] # convert the YYYYMMDD format to a real date
+        activity_date = max(submission_date - timedelta(days=7), min(submission_date, ping["activity_date"])) # normalize the activity date if it's out of range
+
+        usage_hours = min(25, (ping["subsession_length"] or 0) / 3600.0) # usage hours, limited to ~25 hours to keep things normalized
+        main_crashes = int(ping["doc_type"] == "crash") # main process crash (is a crash ping)
+        content_crashes = ping["subprocess_crash_content"] or 0
+        plugin_crashes = ping["subprocess_crash_plugin"] or 0
+        gecko_media_plugin_crashes = ping["subprocess_crash_gmplugin"] or 0
+        return (
+            (submission_date, activity_date) + tuple(ping[key] for key in dimension_names), # all the dimensions we can compare by
+            [ # the crash values
+                1, # number of pings represented by the aggregate
+                usage_hours, main_crashes, content_crashes, plugin_crashes, gecko_media_plugin_crashes,
+
+                # squared versions in order to compute stddev (with $$\sigma = \sqrt{\frac{\sum X^2}{N} - \mu^2}$$)
+                usage_hours ** 2, main_crashes ** 2, content_crashes ** 2, plugin_crashes ** 2, gecko_media_plugin_crashes ** 2,
+            ]
+        )
+
+    crash_values = ping_properties.map(get_crash_pair).reduceByKey(lambda a, b: [x + y for x, y in zip(a, b)])
```

Comment:
Using zip is slow, you might get better perf without it; (you might want to benchmark it).